### ISSUE TITLE
Deprecate `tls::*_file` in favor of `tls::*_pem`

### ISSUE THREAD

**Issue by mx-psi** [open-telemetry/opentelemetry-collector#7691]
After #7676 is merged, we will have duplicate functionality on the `TLSConfig` struct, where  you can pass certificate and keys by providing either a file path (e.g. `tls::ca_file`) or in-memory (e.g. `tls::ca_pem`). Because of our support for `${file:/path/to/file}` URIs, the former is redundant; one can put `tls::ca_pem: ${file:/path/to/ca}` to get the same functionality.

Once #7676 is released, we should deprecate the `tls::*_file` options in favor of `tls::*_pem` options.

**Comment by bogdandrutu** [comment]
@erikbaranowski do you want to fix this?

**Comment by mx-psi** [comment]
No longer on hold

**Comment by jpkrohling** [comment]
[Jaeger Remote Sampling ](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/jaegerremotesampling) as well

**Comment by atoulme** [comment]
I see that reload is implemented for some of the configtls settings but only for the client CA file under the TLS server settings. Is there a use case for live reload of certs and pem files that must be provided by the package, or is the file provider able to provide this somehow?

**Comment by jpkrohling** [comment]
Reloading certs is a VERY common use case. If the file provider can't provide this, we shouldn't remove the _file options.

**Comment by evan-bradley** [comment]
The file provider could be made to reload the certs, but the semantics would be very different from how configtls does this today with the `*_file` options. The `*_file` options allow a running server to reload a certificate by taking advantage of the [`GetCertificate`](https://pkg.go.dev/crypto/tls#Config.GetCertificate) field in `tls.Config` to re-read the file from disk. If the file provider reloaded the certificate, it would force a reload of the whole Collector config and trigger the Collector to restart the service.

I think we will need to keep the `*_file` options at least for the foreseeable future, if not indefinitely.

**Comment by mx-psi** [comment]
Reloading of a single field in the configuration definitely seems difficult to achieve without a significant refactor, I think we can close this as not planned (though I would be interested in hearing @bogdandrutu's thoughts on this).

We can file a separate issue for exploring supporting this reloading (out of scope for 1.0). 

I found a few other components that ask for a filepath in the configuration:
- [jaegerremotesamplingextension](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/extension/jaegerremotesampling)
- [f5cloudexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/f5cloudexporter)
- [kafkaexporter](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/exporter/kafkaexporter)
- [sshcheckreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/sshcheckreceiver)

I can file another issue on contrib to see if we should remove/replace the option on these. Does this make sense?

**Comment by yurishkuro** [comment]
I am surprised the discussion is going in this direction. What percentage of users do we expect to be able to provide data via files vs. in-memory mechanism? By my scientific napkin math it's 1000:1, with 95% confidence.

**Comment by mx-psi** [comment]
@yurishkuro I am waiting on your reply to https://github.com/open-telemetry/opentelemetry-collector/issues/7691#issuecomment-1910534594 to close this issue. Could you have a look?

**Comment by yurishkuro** [comment]
> why should this ratio make us not close this issue?

Because, if you don't dispute the "math", it will break 1000 users per 1 who needs the new syntax?

**Comment by mx-psi** [comment]
> Because, if you don't dispute the "math", it will break 1000 users per 1 who needs the new syntax?

To be clear, we are not going to do any changes on `configtls`, since I believe your comment is about the other components, I am going to close this one as wontfix.

We may change things in other components, following the rule that Evan proposed on https://github.com/open-telemetry/opentelemetry-collector/issues/7691#issuecomment-1910630663. This is the kind of change for which we can provide automated migrations on the Operator and using a tool like #7631 if needed, so I think the breakage would be minor. In any case, we can discuss this in the specific issues for the components
